### ISSUE TITLE
As podcaster I want to create and edit chapter marks manually

### ISSUE THREAD

**Issue by Fischaela** [podlove/radiator-cms#46]

**Comment by Letty** [comment]
Workflow:

1.  Upload file to server
  -> Parse file on server
2. <- Response from server
  Display content on audio timeline with labels / pins for each chapter (moveable)
  Display a sortable table with start point, chapter title and duration (editable, sortable)
3. Add a missing chapter if needed

Content will always be saved on server, no "save" button needed

**Comment by Letty** [comment]
![1024_Episode--new Chaptermarks step 1](https://user-images.githubusercontent.com/465678/61367383-c75a1180-a88b-11e9-9d2d-d9cadf21264e.png)
![1024_Episode--new Chaptermarks step 2](https://user-images.githubusercontent.com/465678/61367384-c75a1180-a88b-11e9-9ce3-36052ce65e5e.png)
![1024_Episode--new Chaptermarks step 3](https://user-images.githubusercontent.com/465678/61367386-c88b3e80-a88b-11e9-8198-da58a7319bfd.png)

**Comment by Fischaela** [comment]
Nice! :)

Regarding the last wireframe: 
* (fyi) the play button cannot always be shown – only if the podcaster has uploaded an audio file yet
* How would one edit the content (strings and timestamps)?
* How would one delete a chapter?

**Comment by eteubert** [comment]
Looking good 👍 

There is no `.md` format for chaptermarks. What we can import so far:

- mp4chaps (.txt)
- Podlove Simple Chapters (.psc)
- Audacity project files
- Hindenburg project files

And just fyi, there are implementations for all of these in JS, see https://www.npmjs.com/search?q=podcast-chapter-parser (but I agree that this should be handled by the API/server in our case).

Not sure a play button is needed in the chapter management UI but the timeline visualization could be useful to see on a quick glance how chapters are distributed as expected.

As @Fischaela mentioned, deleting/editing still needs UI. Podlove Publisher slides an editing sidebar in when a chapter row is selected.

![chapter-editing](https://user-images.githubusercontent.com/235918/61369743-122a5800-a891-11e9-93ab-2ab9421d68eb.gif)

**Comment by Letty** [comment]
Thx!
As @Fischaela mentioned, the delete button is added and next to it edit and sort.
When the user interacts with the edit button the row turns into fields, like in the example from @eteubert 
(The color is just used due to the lack of sketch skill for input fields, i used the old style of inputs)

![1024_Episode--new Chaptermarks step 3_3](https://user-images.githubusercontent.com/465678/61383837-bae7b000-a8af-11e9-923e-ca050a231bea.png)
![1024_Episode--new Chaptermarks step 3_3 editmode](https://user-images.githubusercontent.com/465678/61383836-bae7b000-a8af-11e9-843d-52f61d2febdb.png)

**Comment by Fischaela** [comment]
yeah nice! for development, we should change the edit button to handle "cancel edit" and "save edit"

**Comment by monkeydom** [comment]
Open questions to me: 
* How do links appear in that list, and how to edit them?
* How is the sort widget supposed to work? Dragging just modifies the time? The sort order is implicit based on the start time. (While I would like dragging around in the timeline above, it is probably to imprecise. Just imagine a 5 hour freakshow)
* What is the icon between edit and order? (If it is supposed to be a trashcan for delete I did not recognize it as such)
* How would one add a chapter image, and how would you display that image in that list?
* Adding a chapter probably makes more sense as a click on the timeline or with some connection to the player so it is added at the play position.

**Comment by monkeydom** [comment]
Small one: "Chapter Marks" two words. Imho also the "Upload Chapter Marks File" button should still exist when chapters are present.

**Comment by Letty** [comment]
> Open questions to me:
> 
>  * How do links appear in that list, and how to edit them?
>  * How would one add a chapter image, and how would you display that image in that list?

What is possible as "content" for chapter marks? Links to other "things" on the internet? One image or multiple? Text? cc @eteubert 

>  * How is the sort widget supposed to work? Dragging just modifies the time? The sort order is implicit based on the start time. (While I would like dragging around in the timeline above, it is probably to imprecise. Just imagine a 5 hour Freakshow)

I would not make the timeline interactive, this element should be more like a visual helper.
Dragging inside the list modifies the start time and therefore the order of chapter marks. I hope this is more user friendly then modifying the start time manually.

>  * What is the icon between edit and order? (If it is supposed to be a trashcan for delete I did not recognize it as such)

It's a trashcan and just for wire frames. Buefy can handle font awesome and material design icons.

>  * Adding a chapter probably makes more sense as a click on the timeline or with some connection to the player so it is added at the play position.

If the timeline is already dense with chapter marks, it would be hard. But the button can have an auto fill for start time, based on the player. This is something we should discuss on the weekend.

**Comment by eteubert** [comment]
> What is possible as "content" for chapter marks? Links to other "things" on the internet? One image or multiple? Text?

A chapter consists of:

- a timestamp (required)
- a plain text chapter name (required): no markup, no line breaks
- one URL/link (optional)
- one image (optional)

**Comment by Letty** [comment]
- [ ] why is a 500 error in the backend when creating or reading chapter marks?
- [ ] transform time in ms into human readable time
- [ ] show a link
- [ ] show an image
